manager/integration/tests/test_scheduling.py

```diff
@@ -454,6 +468,424 @@ def test_data_locality_basic():
+        replica with Spec.HardNodeAffinity=node-3
+    40. Increase numberOfReplica to 3. Verify that the replica set contains:
+        one on node-1, one on node-2,  one failed replica
+        with Spec.HardNodeAffinity=node-3.
+    41. Decrease numberOfReplica to 2. Verify that the replica set contains:
+        one on node-1, one on node-2,  one failed replica
+        with Spec.HardNodeAffinity=node-3.
+    42. Decrease numberOfReplica to 1. Verify that the replica set contains:
+        one on node-1 or node-2,  one failed replica
+        with Spec.HardNodeAffinity=node-3.
+    43. Decrease numberOfReplica to 2. Verify that the replica set contains:
+        one on node-1, one on node-2, one failed replica
+        with Spec.HardNodeAffinity=node-3.
+    44. Turn off data locality by set `dataLocality=disabled` for the vol.
+        Verify that the replica set contains: one on node-1, one on node-2
+
+    45. clean up
     """
-    pass
+    nodes = client.list_node()
+
+    default_data_locality_setting = \
+        client.by_id_setting(SETTING_DEFAULT_DATA_LOCALITY)
+    try:
+        client.update(default_data_locality_setting, value="disabled")
+    except Exception as e:
+        print("Exception when update Default Data Locality setting",
+              default_data_locality_setting, e)
+
+    volume1_name = volume_name + "-1"
+    volume1_size = str(500 * Mi)
+    volume1_data_path = "/data/test"
+    pv1_name = volume1_name + "-pv"
+    pvc1_name = volume1_name + "-pvc"
+    pod1_name = volume1_name + "-pod"
+    pod1 = pod
+
+    pod1['metadata']['name'] = pod1_name
+
+    volume1 = create_and_check_volume(client,
+                                      volume1_name,
+                                      num_of_replicas=1,
+                                      size=volume1_size)
+
+    volume1 = client.by_id_volume(volume1_name)
+    create_pv_for_volume(client, core_api, volume1, pv1_name)
+    create_pvc_for_volume(client, core_api, volume1, pvc1_name)
+
+    volume1 = client.by_id_volume(volume1_name)
+    volume1_replica_node = volume1.replicas[0]['hostId']
+
+    volume1_attached_node = None
+    for node in nodes:
+        if node.name != volume1_replica_node:
+            volume1_attached_node = node.name
+            break
+
+    assert volume1_attached_node is not None
+
+    pod1['spec']['volumes'] = [{
+        "name": "pod-data",
+        "persistentVolumeClaim": {
+            "claimName": pvc1_name
+        }
+    }]
+
+    pod1['spec']['nodeSelector'] = \
+        {"kubernetes.io/hostname": volume1_attached_node}
+    create_and_wait_pod(core_api, pod1)
+
+    write_pod_volume_random_data(core_api, pod1_name,
+                                 volume1_data_path, DATA_SIZE_IN_MB_2)
+
+    for i in range(10):
+        volume1 = client.by_id_volume(volume1_name)
+        assert len(volume1.replicas) == 1
+        assert volume1.replicas[0]['hostId'] == volume1_replica_node
```

Comment:
I guess that we are imeplementing step #5  here. Should we do like `assert volume1.replicas[0]['hostId'] != volume1_attached_node` so it is a bit clearer that the replica is not on the same node as the engine. It is totally fine not to change though. 🙂

Reply:
fixed